### ISSUE TITLE
10.4 LeNet-5实战 loss值更新与笔记里更新方式不同

### ISSUE THREAD

**Issue by Sun-Lixin** [dragen1860/Deep-Learning-with-TensorFlow-book#184]
这是书上写的

```python
# 创建损失函数的类，在实际计算时直接调用类实例即可 
criteon = losses.CategoricalCrossentropy(from_logits=True) 
训练部分实现如下： 
    # 构建梯度记录环境 
    with tf.GradientTape() as tape:  
        # 插入通道维度，=>[b,28,28,1] 
        x = tf.expand_dims(x,axis=3)
        # 前向计算，获得 10 类别的概率分布，[b, 784] => [b, 10] 
		out = network(x) 
        # 真实标签 one-hot 编码，[b] => [b, 10] 
		y_onehot = tf.one_hot(y, depth=10) 
        # 计算交叉熵损失函数，标量 
		loss = criteon(y_onehot, out) 
    # 自动计算梯度 
    grads = tape.gradient(loss, network.trainable_variables) 
    # 自动更新参数 
    optimizer.apply_gradients(zip(grads, 					network.trainable_variables)) 
```

这里是 loss = criteon(...)
但是与作者介绍的那位童鞋的笔记不太一样

```python
def train(train_db, network, criteon, optimizer, epoch_num):
    for epoch in range(epoch_num):
        correct, total, loss = 0, 0, 0
        for step, (x, y) in enumerate(train_db):
            # 构建梯度记录环境
            with tf.GradientTape() as tape:
                # 插入通道维度， =>[b,28,28,1]
                x = tf.expand_dims(x, axis=3)
                # 前向计算，获得 10 类别的概率分布， [b, 784] => [b, 10]
                out = network(x)
                pred = tf.argmax(out, axis=-1)
                # 真实标签 one-hot 编码， [b] => [b, 10]
                y_onehot = tf.one_hot(y, depth=10)
                # 计算交叉熵损失函数，标量
                loss += criteon(y_onehot, out)
                # 统计预测正确数量
                correct += float(tf.reduce_sum(tf.cast(tf.equal(pred, y), tf.float32)))
                # 统计预测样本总数
                total += x.shape[0]


            # 自动计算梯度
            grads = tape.gradient(loss, network.trainable_variables)
            # 自动更新参数
            optimizer.apply_gradients(zip(grads, network.trainable_variables))

        print(epoch, 'loss=', float(loss), 'acc=', correct / total)

    return network
```

这里loss是 += criteon(...)

在后面 tape.gradient(loss, ...)使用到loss会有影响，所以我不太明白，到底是哪个地方有问题。

**Comment by Sun-Lixin** [comment]
还有图2中的acc更高

**Comment by dragen1860** [comment]
建议按书本中的来。稍后会修复此处代码。
